source/simulator/helper_functions.cc

```diff
@@ -1529,13 +1529,39 @@ namespace aspect
 
 
 
+  template <int dim>
+  void Simulator<dim>::update_solution_vectors_with_reaction_results (const unsigned int block_index,
+                                                                      const LinearAlgebra::BlockVector &distributed_vector,
+                                                                      const LinearAlgebra::BlockVector &distributed_reaction_vector)
+  {
+    solution.block(block_index) = distributed_vector.block(block_index);
+
+    // we have to update the old solution with our reaction update too
+    // so that the advection scheme will have the correct time stepping in the next step
+    LinearAlgebra::BlockVector tmp(distributed_vector);
+
+    tmp.block(block_index) = old_solution.block(block_index);
+    tmp.block(block_index) +=  distributed_reaction_vector.block(block_index);
+    old_solution.block(block_index) = tmp.block(block_index);
```

Comment:
So if I see this right, you're doing
```
  A = B
  A += C
  B = A
```
This would be equivalent to `B += C`. Why does that not work? Because the distribution of vectors is different? If that's the case, could you add a comment of the form
```
  // What we really want to do is 
  //     old_solution.block(block_index) += distributed_reaction_vector.block(block_index);
  // but because 'old_solution' is a ghosted vector, we can't write into it directory. Rather,
  // we have to go around with a completely distributed vector.
```